File: src/main/java/mediathek/tool/models/TModelFilm.java
package mediathek.tool.models;

import mediathek.daten.DatenFilm;
import mediathek.tool.FilmSize;
import mediathek.tool.datum.DatumFilm;

import javax.swing.table.AbstractTableModel;
import java.util.ArrayList;
import java.util.List;

public class TModelFilm extends AbstractTableModel {
    private static final int COLUMN_COUNT = 15;
    private final List<DatenFilm> dataList;

    public TModelFilm() {
        dataList = new ArrayList<>();
    }

    public TModelFilm(int capacity) {
        dataList = new ArrayList<>(capacity);
    }

    @Override
    public boolean isCellEditable(int i, int j) {
        return false;
    }

    /**
     * Get model row for a requested film nr.
     *
     * @param reqFilmNr the filmnr of the request
     * @return the model row, otherwise -1
     */
    public int getModelRowForFilmNumber(int reqFilmNr) {
        int ret = 0;

        for (var film : dataList) {
            if (film.getFilmNr() == reqFilmNr)
                return ret;
            else
                ret++;
        }

        return -1;
    }

    @Override
    public int getRowCount() {
        return dataList.size();
    }

    @Override
    public int getColumnCount() {
        return COLUMN_COUNT;
    }

    @Override
    public Class<?> getColumnClass(int columnIndex) {
        return switch (columnIndex) {
            case DatenFilm.FILM_NR, DatenFilm.FILM_DAUER -> Integer.class;
            case DatenFilm.FILM_DATUM -> DatumFilm.class;
            case DatenFilm.FILM_GROESSE -> FilmSize.class;
            case DatenFilm.FILM_HD, DatenFilm.FILM_UT -> Boolean.class;
            case DatenFilm.FILM_DATUM_LONG -> Long.class;
            default -> String.class;
        };
    }

    @Override
    public String getColumnName(int column) {
        return switch (column) {
            case DatenFilm.FILM_ABSPIELEN, DatenFilm.FILM_AUFZEICHNEN, DatenFilm.FILM_MERKEN -> "";
            case DatenFilm.FILM_NR -> "Nr";
            case DatenFilm.FILM_SENDER -> "Sender";
            case DatenFilm.FILM_THEMA -> "Thema";
            case DatenFilm.FILM_TITEL -> "Titel";
            case DatenFilm.FILM_DATUM -> "Datum";
            case DatenFilm.FILM_ZEIT -> "Zeit";
            case DatenFilm.FILM_DAUER -> "Dauer";
            case DatenFilm.FILM_GROESSE -> "Größe [MB]";
            case DatenFilm.FILM_HD -> "HQ";
            case DatenFilm.FILM_UT -> "UT";
            case DatenFilm.FILM_GEO -> "Geo";
            case DatenFilm.FILM_URL -> "URL";
            default -> throw new IndexOutOfBoundsException("UNKNOWN COLUMN NAME: " + column);
        };
    }

    @Override
    public Object getValueAt(int row, int column) {
        final var film = dataList.get(row);

        return switch (column) {
            case DatenFilm.FILM_NR -> film.getFilmNr();
            case DatenFilm.FILM_SENDER -> film.getSender();
            case DatenFilm.FILM_THEMA -> film.getThema();
            case DatenFilm.FILM_TITEL -> film.getTitle();
            case DatenFilm.FILM_ABSPIELEN, DatenFilm.FILM_AUFZEICHNEN, DatenFilm.FILM_MERKEN -> "";
            case DatenFilm.FILM_DATUM -> film.getDatumFilm();
            case DatenFilm.FILM_ZEIT -> film.getSendeZeit();
            case DatenFilm.FILM_DAUER -> film.getDuration();
            case DatenFilm.FILM_GROESSE -> film.getFilmSize();
            case DatenFilm.FILM_HD -> film.isHighQuality();
            case DatenFilm.FILM_UT -> film.hasSubtitle();
            case DatenFilm.FILM_GEO -> film.getGeo().orElse("");
            case DatenFilm.FILM_URL -> film.getUrl();
            case DatenFilm.FILM_REF -> film;
            default -> throw new IndexOutOfBoundsException("UNKNOWN COLUMN VALUE: " + column);
        };
    }

    public void addAll(List<DatenFilm> listeFilme) {
        final int oldRowCount = dataList.size();
        dataList.addAll(listeFilme);
        fireTableRowsInserted(oldRowCount, dataList.size());
    }

    public void addRow(DatenFilm film) {
        dataList.add(film);
        final int rowCount = dataList.size();
        fireTableRowsInserted(rowCount, rowCount);
    }
}

File: src/main/kotlin/mediathek/tool/models/TModelFilm.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.tool.models

import mediathek.daten.DatenFilm
import javax.swing.table.AbstractTableModel

class TModelFilm(capacity: Int = 0) : AbstractTableModel() {
    private val dataList = ArrayList<DatenFilm>(capacity)

    override fun getRowCount(): Int = dataList.size

    override fun getColumnCount(): Int = FilmColumn.TABLE_COLUMN_COUNT

    override fun getColumnClass(columnIndex: Int): Class<*> =
        FilmColumn.fromIndex(columnIndex).valueType

    override fun getColumnName(column: Int): String =
        FilmColumn.fromIndex(column).title()

    override fun getValueAt(row: Int, column: Int): Any {
        val film = dataList[row]
        return FilmColumn.fromIndex(column).valueFrom(film)
    }

    fun addAll(listeFilme: List<DatenFilm>) {
        if (listeFilme.isEmpty()) {
            return
        }
        val oldRowCount = dataList.size
        dataList.addAll(listeFilme)
        fireTableRowsInserted(oldRowCount, dataList.lastIndex)
    }
}
